### ISSUE TITLE
additional small changes/qs re sync

### ISSUE THREAD

**Issue by phritz** [rocicorp/repc#107]
I looked over what we need to implement sync beyond what is represented in issues already and came up with the small grab bag of issues/questions below, might as well discuss up front.

1. We need to do a lot of traversing chains of commits for sync. This requires reading commits that are not at the head of their chain. I plan to make some small tweaks to the commit and tx interfaces to make this easier. Right now functions like `Commit::from_head` bundle two things together, getting the hash the named head is set to and then reading it into a Commit. I'm going to make at least the reading into a commit into a first class citizen. Something like the following, presumably this is uncontroversial:
   - let's have "load" mean "read from the store" so rename `Commit::load()` to `Commit::from_chunk()`.
   - add `Commit::load(hash, dag_read) -> Commit` that calls `dag_read.get_chunk(hash)` and then `Commit::from_chunk` on the result
   - have `from_head` call `dag_read.get_head` as it does today then `Commit::load`
1. Do we use multiple sync branches, allowing multiple syncs to run in parallel? IF enabling parallel syncs is important then i think we should make a small change to the protocol (see below) even though we can technically do it without the change. IF enabling parallel syncs is not important then it's even easier (see also below).
   - with no changes to the protocol the most straightforward thing is to not allows syncs in parallel. If we use a single named head to track the sync chain then if a new sync starts while one is already running it effectively cancels the old sync. (i can go into details if you want, not sure they are important.)
   - with no changes to the protocol we can support multiple parallel syncs with a creative use of named heads (a sync chain has a named head with name equal to the hash of its head commit; to extend a sync chain with a commit delete the old named head and crate a new named head with name hash of the new commit. sync chains wont be gc'd and we always have synchead/basis passed in in open/committransaction and maybeendsync.)
   - if we added syncid to the rebaseopts then we can enable multiple parallel syncs using a named head equal to the syncid. 
1. We need to add a way to remove a named head to clean up after sync. This seems straightforward.
1. We need to fix this bug https://github.com/rocicorp/repc/blob/e803f9a3b38da5bcce0112b3436d746ded9200b4/src/embed/connection.rs#L182 (basis is a hash, not a named head).

**Comment by aboodman** [comment]
On Tue, Aug 18, 2020 at 11:57 AM Phritz <notifications@github.com> wrote:

> I looked over what we need to implement sync beyond what is represented in
> issues already and came up with the small grab bag of issues/questions
> below, might as well discuss up front.
>
>    1. We need to do a lot of traversing chains of commits for sync. This
>    requires reading commits that are not at the head of their chain. I plan to
>    make some small tweaks to the commit and tx interfaces to make this easier.
>    Right now functions like Commit::from_head bundle two things together,
>    getting the hash the named head is set to and then reading it into a
>    Commit. I'm going to make at least the reading into a commit into a first
>    class citizen. Something like the following, presumably this is
>    uncontroversial:
>       - let's have "load" mean "read from the store" so rename
>       Commit::load() to Commit::from_chunk().
>       - add Commit::load(hash, dag_read) -> Commit that calls
>       dag_read.get_chunk(hash) and then Commit::from_chunk on the result
>       - have from_head call dag_read.get_head as it does today then
>       Commit::load
>
> I have this change pending in my client too. I'll try and split it out and
land it tonight.

>
>    1.
>       2. Do we use multiple sync branches, allowing multiple syncs to run
>    in parallel? IF enabling parallel syncs is important then i think we should
>    make a small change to the protocol (see below) even though we can
>    technically do it without the change. IF enabling parallel syncs is not
>    important then it's even easier (see also below).
>       - with no changes to the protocol the most straightforward thing is
>       to not allows syncs in parallel. If we use a single named head to track the
>       sync chain then if a new sync starts while one is already running it
>       effectively cancels the old sync. (i can go into details if you want, not
>       sure they are important.)
>       - with no changes to the protocol we can support multiple parallel
>       syncs with a creative use of named heads (a sync chain has a named head
>       with name equal to the hash of its head commit; to extend a sync chain with
>       a commit delete the old named head and crate a new named head with name
>       hash of the new commit. sync chains wont be gc'd and we always have
>       synchead/basis passed in in open/committransaction and maybeendsync.
>
>
>    1.
>       - if we added syncid to the rebaseopts then we can enable multiple
>       parallel syncs using a named head equal to the syncid.
>
> I like this the first and third options. I guess minor preference for the
first one since it's "less".

>
>    1.
>       2. We need to add a way to remove a named head to clean up after
>    sync. This seems straightforward.
>    3. We need to fix this bug
>    https://github.com/rocicorp/repc/blob/e803f9a3b38da5bcce0112b3436d746ded9200b4/src/embed/connection.rs#L182
>    (basis is a hash, not a named head).
>
> Whoops I was going to add more extensive tests for this area once the API
was shaped better.

**Comment by phritz** [comment]
thanks. no rush on landing your changes. i'm getting a late start tomorrow afternoon.
